### PR TITLE
Support removing TLS handlers from live connections.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

In some unusual protocol scenarios it is desirable to remove a TLS
tunnel from an active connection. Currently NIOOpenSSL does not
support doing this: the only way to get it to issue a CLOSE_NOTIFY
is to call close() on the Channel. This is less than ideal.

This patch will continue to fail tests until apple/swift-nio#651 lands in a release. At that time, I'll update this repo to require the new SwiftNIO release, and everything should be hunky-dory.

Modifications:

- Added a OpenSSLHandler.stopTLS method
- Reworked the internal state machine to handle this use-case.

Result:

Users can safely unwrap TLS from a Channel.

### CODE REVIEW

**Comment by Lukasa** [comment]
@weissi ready for review

File: Tests/NIOOpenSSLTests/UnwrappingTests.swift [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+    func testSimpleUnwrapping() throws {
+        let serverChannel = EmbeddedChannel()
+        let clientChannel = EmbeddedChannel()
+
+        var clientClosed = false
+        var serverClosed = false
+        var unwrapped = false
+
+        defer {
+            // We expect the server case to throw
```
Comment (weissi) [inline Tests/NIOOpenSSLTests/UnwrappingTests.swift]: can we assert that then with `XCTAssertThrows`?

File: Tests/NIOOpenSSLTests/UnwrappingTests.swift [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+            _ = try? serverChannel.finish()
```
Comment (weissi) [inline Tests/NIOOpenSSLTests/UnwrappingTests.swift]: assert more stuff?

File: Tests/NIOOpenSSLTests/UnwrappingTests.swift [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+    func testUnwrappingFollowedByClosure() throws {
+        let serverChannel = EmbeddedChannel()
+        let clientChannel = EmbeddedChannel()
+
+        var clientClosed = false
+        var clientUnwrapped = false
+
+        defer {
+            // Both channels will already be closed
+            _ = try? serverChannel.finish()
```
Comment (weissi) [inline Tests/NIOOpenSSLTests/UnwrappingTests.swift]: in which case we can just assert `XCTAssertFalse(serverChannel.isActive)` no?

File: Tests/NIOOpenSSLTests/UnwrappingTests.swift [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+    func testUnwrappingMeetsTCPFIN() throws {
+        let serverChannel = EmbeddedChannel()
+        let clientChannel = EmbeddedChannel()
+
+        var clientClosed = false
+        var clientUnwrapped = false
+
+        defer {
+            // The errors here are expected
+            _ = try? serverChannel.finish()
```
Comment (weissi) [inline Tests/NIOOpenSSLTests/UnwrappingTests.swift]: asserts for that?

File: Tests/NIOOpenSSLTests/UnwrappingTests.swift [hunk 5]
```diff
@@ -0,0 +1,750 @@
...
+
+    func testDoubleUnwrapping() throws {
+        let serverChannel = EmbeddedChannel()
+        let clientChannel = EmbeddedChannel()
+
+        var promiseCalled = false
+
+        defer {
+            // We expect the server case to throw
+            _ = try? serverChannel.finish()
```
Comment (weissi) [inline Tests/NIOOpenSSLTests/UnwrappingTests.swift]: asserts for that too?

**Pull request merged**